File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/observability/Observability.kt
package app.pantopus.android.data.observability

import android.content.Context
import app.pantopus.android.BuildConfig
import io.sentry.Breadcrumb
import io.sentry.Sentry
import io.sentry.SentryLevel
import io.sentry.android.core.SentryAndroid
import io.sentry.android.timber.SentryTimberIntegration
import io.sentry.protocol.User
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Single entry point for crash reporting, error capture, and analytics.
 *
 * Call [start] from [app.pantopus.android.PantopusApplication.onCreate]; every other
 * piece of the app goes through this class so the underlying backend (Sentry, Firebase, …)
 * can be swapped without touching call sites.
 *
 * DSN comes from `BuildConfig.SENTRY_DSN` — if empty, Sentry is not initialised and
 * everything no-ops (useful for local dev and CI).
 */
@Singleton
class Observability
    @Inject
    constructor() {
        private var started = false

        fun start(context: Context) {
            if (started) return
            started = true

            val dsn = BuildConfig.SENTRY_DSN
            if (dsn.isBlank()) {
                Timber.i("Sentry DSN not set — crash reporting disabled")
                return
            }

            SentryAndroid.init(context) { options ->
                options.dsn = dsn
                options.environment = BuildConfig.PANTOPUS_ENV
                options.release = "app.pantopus.android@${BuildConfig.VERSION_NAME}+${BuildConfig.VERSION_CODE}"
                options.isEnableAutoSessionTracking = true
                options.tracesSampleRate = if (BuildConfig.PANTOPUS_ENV == "production") 0.1 else 1.0
                options.isAttachScreenshot = false
                options.isAttachViewHierarchy = false
                options.isSendDefaultPii = false
                // Route Timber.e/w into Sentry breadcrumbs + events automatically.
                options.addIntegration(
                    SentryTimberIntegration(
                        minEventLevel = SentryLevel.ERROR,
                        minBreadcrumbLevel = SentryLevel.INFO,
                    ),
                )
            }
            Timber.i("Sentry started (env=${BuildConfig.PANTOPUS_ENV})")
        }

        fun capture(throwable: Throwable) {
            Timber.e(throwable)
            if (started) Sentry.captureException(throwable)
        }

        fun capture(
            message: String,
            level: SentryLevel = SentryLevel.WARNING,
        ) {
            Timber.log(level.toTimberPriority(), message)
            if (started) Sentry.captureMessage(message, level)
        }

        fun identify(
            userId: String?,
            email: String? = null,
        ) {
            if (!started) return
            if (userId == null) {
                Sentry.setUser(null)
            } else {
                Sentry.setUser(
                    User().apply {
                        this.id = userId
                        this.email = email
                    },
                )
            }
        }

        /** Lightweight analytics — structured log + Sentry breadcrumb. */
        fun track(
            name: String,
            properties: Map<String, String> = emptyMap(),
        ) {
            Timber.i("analytics event=$name props=$properties")
            if (!started) return
            val crumb =
                Breadcrumb().apply {
                    category = "analytics"
                    message = name
                    level = SentryLevel.INFO
                    properties.forEach { (k, v) -> setData(k, v) }
                }
            Sentry.addBreadcrumb(crumb)
        }
    }

private fun SentryLevel.toTimberPriority(): Int =
    when (this) {
        SentryLevel.DEBUG -> android.util.Log.DEBUG
        SentryLevel.INFO -> android.util.Log.INFO
        SentryLevel.WARNING -> android.util.Log.WARN
        SentryLevel.ERROR, SentryLevel.FATAL -> android.util.Log.ERROR
    }

File: frontend/apps/ios/Pantopus/Core/Observability/Observability.swift
//
//  Observability.swift
//  Pantopus
//
//  Single entry point for crash reporting, error capture, and analytics.
//  Everything else in the app goes through `Observability.shared` so we can
//  swap the backend (Sentry, Firebase, …) without touching call sites.
//

import Foundation
import Logging
import Sentry

@MainActor
final class Observability {
    static let shared = Observability()

    private let logger = Logger(label: "app.pantopus.ios.Observability")
    private(set) var isStarted = false

    private init() {}

    /// Call once from `AppDelegate.didFinishLaunching`. Safe to call again — no-ops after the first invocation.
    func start(environment: AppEnvironment) {
        guard !isStarted else { return }
        isStarted = true

        let dsn = Bundle.main.object(forInfoDictionaryKey: "SentryDSN") as? String ?? ""
        guard !dsn.isEmpty else {
            logger.info("Sentry DSN not set — skipping crash reporting init")
            return
        }

        SentrySDK.start { options in
            options.dsn = dsn
            options.environment = environment.target.rawValue
            options.releaseName = Self.releaseName
            options.enableAutoSessionTracking = true
            options.enableAutoPerformanceTracing = true
            options.tracesSampleRate = environment.target == .production ? 0.1 : 1.0
            options.enableUserInteractionTracing = true
            options.enableNetworkTracking = true
            options.attachViewHierarchy = false
            options.attachScreenshot = false
            options.sendDefaultPii = false
            options.beforeSend = { event in
                // Drop noisy low-signal events here if needed.
                event
            }
        }

        logger.info("Sentry started", metadata: ["env": .string(environment.target.rawValue)])
    }

    // MARK: - Error capture

    func capture(_ error: Error, file: StaticString = #fileID, line: UInt = #line) {
        logger.error("\(error)", metadata: ["file": .string("\(file):\(line)")])
        guard isStarted else { return }
        SentrySDK.capture(error: error)
    }

    func capture(message: String, level: SentryLevel = .warning) {
        logger.warning("\(message)")
        guard isStarted else { return }
        SentrySDK.capture(message: message) { scope in
            scope.setLevel(level)
        }
    }

    // MARK: - User context

    func identify(userId: String?, email: String? = nil) {
        guard isStarted else { return }
        if let userId {
            let user = User(userId: userId)
            user.email = email
            SentrySDK.setUser(user)
        } else {
            SentrySDK.setUser(nil)
        }
    }

    // MARK: - Breadcrumbs & analytics

    /// Lightweight analytics event. For now this is a structured log + Sentry breadcrumb;
    /// wire in a real analytics SDK (Amplitude / Mixpanel / PostHog) by extending this method.
    func track(_ name: String, properties: [String: String] = [:]) {
        logger.info("analytics", metadata: [
            "event": .string(name),
            "props": .string(properties.description),
        ])
        guard isStarted else { return }
        let crumb = Breadcrumb(level: .info, category: "analytics")
        crumb.message = name
        crumb.data = properties
        SentrySDK.addBreadcrumb(crumb)
    }

    // MARK: - Release tag

    private static var releaseName: String {
        let version = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0.0"
        let build = Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "0"
        return "app.pantopus.ios@\(version)+\(build)"
    }
}
